### DISCUSSION TITLE
Can't get column type inference to work in v2

### DISCUSSION THREAD

**Discussion by lukasjuhrich** [sqlalchemy/sqlalchemy#8127] [Usage Questions] [answered]
Hi,

I was trying to see how the „builtin“ typing worked in sqlalchemy2.0 and tested the current `HEAD` (b163099ada4009b2bdc9538c04120aee1f13dcc8) with `mypy==0.961` against the following snippet:

```python
from sqlalchemy import Column
from sqlalchemy.orm import registry, as_declarative
from sqlalchemy.orm.decl_api import DeclarativeMeta
from sqlalchemy.types import Integer


mapper_registry = registry()

# Yes, that's on purpose, as I have a nontrivial metaclass in production
class MyMeta(DeclarativeMeta): pass


@as_declarative(metaclass=MyMeta)
class ModelBase:
    __abstract__ = True
    # static attribute definitions for mypy
    registry = mapper_registry
    metadata = mapper_registry.metadata

    __init__ = mapper_registry.constructor


class Stub(ModelBase):
    id = Column(Integer, primary_key=True)


if __name__ == "__main__":
    import typing

    stub = Stub(id=5)
    if typing.TYPE_CHECKING:
        reveal_type(stub.id)

    assert stub.id.startswith("foo")
```

Clearly, `stub.id` should be `int` and the `startswith` call should fail to type check.
Nonetheless, `mypy --strict test.py` reports no problems, and actually infers `Any`:

```
(sqla20) ➜  sqla2_type_inference git:(master) mypy --strict test.py 
test.py:33: note: Revealed type is "Any"
Success: no issues found in 1 source file
```
Am I doing the base class setup wrong?

See https://github.com/lukasjuhrich/sqla20_type_inference/ for easy reproduction.

**Comment by zzzeek** [comment, marked as answer]
> Hi,
> 
> I was trying to see how the „builtin“ typing worked in sqlalchemy2.0 and tested the current `HEAD` ([b163099](https://github.com/sqlalchemy/sqlalchemy/commit/b163099ada4009b2bdc9538c04120aee1f13dcc8)) with `mypy==0.961` against the following snippet:
> 
> ```python
> from sqlalchemy import Column
> from sqlalchemy.orm import registry, as_declarative
> from sqlalchemy.orm.decl_api import DeclarativeMeta
> from sqlalchemy.types import Integer
> 
> 
> mapper_registry = registry()
> 
> # Yes, that's on purpose, as I have a nontrivial metaclass in production
> class MyMeta(DeclarativeMeta): pass
> 
> 
> @as_declarative(metaclass=MyMeta)
> class ModelBase:
>     __abstract__ = True
>     # static attribute definitions for mypy
>     registry = mapper_registry
>     metadata = mapper_registry.metadata
> 
>     __init__ = mapper_registry.constructor
> 
> 
> class Stub(ModelBase):
>     id = Column(Integer, primary_key=True)

So that's not the new "builtin" typing.  The new typing has not been documented yet, so unless you are looking at the test cases or reading some of the tweets I made, or maybe you found [mapped_column](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html?highlight=mapped_column#sqlalchemy.orm.mapped_column), you wouldn't know how to use the new features.

>     if typing.TYPE_CHECKING:
>         reveal_type(stub.id)
> 
>     assert stub.id.startswith("foo")
> ```
> 
> Clearly, `stub.id` should be `int` and the `startswith` call should fail to type check. 

That code is working as it would be expected to, "id" is a Column object, the datatype is ``Column[int]`` and it has a startswith() method.    Since we assume you arent using the [plugin](https://docs.sqlalchemy.org/en/14/orm/extensions/mypy.html) that would make Mypy see those Column objects as descriptors, there's no way Mypy would know that the ORM is going to replace the Column object with a descriptor.

for examples of the new typing see the test cases at https://github.com/sqlalchemy/sqlalchemy/blob/main/test/orm/declarative/test_typed_mapping.py

**Reply by lukasjuhrich** [reply]
`mapped_column` – together with using [`DeclarativeBase`](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.DeclarativeBase) allows me to make it work.

Thanks for your quick reaction and the enormous amount of work on the typing infrastructure!

**Reply by zzzeek** [reply]
I've pushed an updated version of the [ORM Quickstart](https://docs.sqlalchemy.org/en/20/orm/quickstart.html) which just shows one mapping.  many more aspects of this have yet to be documented.
